'this' in case statement causes compilation error

```scala
class A {
  this match {
    case this => println("foo")
    case _ => ???
   }
}
```

```
A.scala:3: error: '.' expected but '=>' found.
    case this => println("foo")
              ^
A.scala:4: error: identifier expected but ';' found.
    case _ => ???
^
two errors found
```

--- a/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
+++ b/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
@@ -1388,7 +1388,7 @@ self =>
      *  }}}
      */
     def stableId(): Tree =
-      path(thisOK = false, typeOK = false)
+      path(thisOK = true, typeOK = false)
 
     /** {{{
     *   QualId ::= Id {`.` Id}
@@ -2246,10 +2246,8 @@ self =>
        *
        * XXX: Hook for IDE
        */
-      def simplePattern(): Tree = (
-        // simple diagnostics for this entry point
-        simplePattern(() => syntaxErrorOrIncompleteAnd("illegal start of simple pattern", skipIt = true)(errorPatternTree))
-      )
+      def simplePattern(): Tree =
+        simplePattern(() => syntaxErrorOrIncompleteAnd("illegal start of simple pattern", skipIt = true)(errorPatternTree)) // simple diagnostics for this entry point
       def simplePattern(onError: () => Tree): Tree = {
         val start = in.offset
         in.token match {

--- a/test/files/pos/t8365.scala
+++ b/test/files/pos/t8365.scala
@@ -1,0 +1,10 @@
+
+class A {
+  def f =
+    this match {
+      case this =>
+    }
+  val this = this
+}
+
+

